### PR TITLE
#22 | Add getOrCreateIndex functionality

### PR DESCRIPTION
**Pull request by arjunrc143**

### CODE REVIEW

**Comment by eskombro** [comment]
Hi @arjunrc143 

Thanks for your PR. It is important that in the `getoOrCreateIndex` method we catch the Exception only if the error corresponds to [an index that exists already](https://docs.meilisearch.com/errors/#index_already_exists). Otherwise we shouldn't proceed as we do not necessarily know where the problem is coming from. You can check in the MeiliSearch response if the errorCode corresponds to `index_already_exists`.

This will be updated later when we implement a Custom error handler in #7 

Can you please add this check?

**Comment by eskombro** [comment]
Also, please link the issue. You can add a comment in your pr like this

```
Closes #22 
```

File: src/main/java/com/meilisearch/sdk/Client.java [hunk 1]
```diff
@@ -107,4 +107,20 @@ public String updateIndex(String uid, String primaryKey) throws Exception {
 	public String deleteIndex(String uid) throws Exception {
 		return this.indexesHandler.delete(uid);
 	}
+
+	/**
+	 * Get single index by uid or if it does not exists, Create index
+	 *
+	 * @param uid        Unique identifier for the index to create
+	 * @param primaryKey The primary key of the documents in that index
+	 * @return Meilisearch API response
```
Comment (eskombro) [inline src/main/java/com/meilisearch/sdk/Client.java:116]: This function returns an index Instance, not the http response :)

**Review by eskombro — CHANGES_REQUESTED** [review]

File: src/main/java/com/meilisearch/sdk/Client.java (hunk 1 above)
Comment (eskombro) [inline src/main/java/com/meilisearch/sdk/Client.java:116]:
```suggestion
	 * @return Index instance
```

Comment (eskombro) [reply src/main/java/com/meilisearch/sdk/Client.java:116]:
```suggestion
	 * @return Index instance
```

**Comment by arjunrc143** [comment]
> Hi @arjunrc143
> 
> Thanks for your PR. It is important that in the `getoOrCreateIndex` method we catch the Exception only if the error corresponds to [an index that exists already](https://docs.meilisearch.com/errors/#index_already_exists). Otherwise we shouldn't proceed as we do not necessarily know where the problem is coming from. You can check in the MeiliSearch response if the errorCode corresponds to `index_already_exists`.
> 
> This will be updated later when we implement a Custom error handler in #7
> 
> Can you please add this check?

Done

**Comment by arjunrc143** [comment]
closes #22

**Comment by eskombro** [comment]
> closes #22

I think this must be done in the description :)

I linked it anyways !

**Comment by eskombro** [comment]
This PR seems good! Just need to add a few integration tests ! For example:

- Use getOrCreateIndex() and check that the index is created
- Use getOrCreateIndex() with an existing index and check that no error is thrown and the index is returned
- Use getOrCreateIndex() 2 or three times in a row...

**Comment by arjunrc143** [comment]
@eskombro I have added the integration tests. Please check.

**Comment by arjunrc143** [comment]
2 integration tests are failing. Will check that.

**Comment by eskombro** [comment]
I checked what's going on, and the problem is that when you get an `index_already_exists` error, the SDK is throwing an IOException with the message 

```
Server returned HTTP response code: 400 for URL: http://localhost:7700/indexes
```

So the check you implemented for the error catching doesn't work. This problem is coming from what `createIndex()` returns in case of failure, and so, not your bad or any propblem in your PR at all actually.

I'll fix this since I'm implementing a basic error handler, and let you know so you can rebase your branch and it should work fine!

Glad we did those tests 👍

**Comment by arjunrc143** [comment]
@eskombro Cool. Let me know once this gets fixed. Thanks!

File: src/main/java/com/meilisearch/sdk/Client.java [hunk 2]
```diff
@@ -107,4 +107,23 @@ public String updateIndex(String uid, String primaryKey) throws Exception {
...
+	 */
+	public Index getOrCreateIndex(String uid, String primaryKey) throws Exception {
+		try {
+			return this.createIndex(uid, primaryKey);
+		} catch (Exception e) {
+			if(e.getMessage().contains("index_already_exists")) {
+				return this.getIndex(uid);
+			}
+			throw e;
+		}
```
Comment (eskombro) [inline src/main/java/com/meilisearch/sdk/Client.java:127]:
```suggestion
		try {
			return this.createIndex(uid, primaryKey);
		} catch (MeiliSearchApiException e) {
			if(e.getErrorCode().equals("index_already_exists")) {
				return this.getIndex(uid);
			}
			throw e;
		}
```

**Review by eskombro — CHANGES_REQUESTED** [review]
Hey!

We can already use custom Exceptions, Can you please rebase your branch on master?

I did a suggestion on how to check for the MeiliSearchApiException in this method which is already tested! 

ALSO:

Can you please create a `getOrCreateIndex(String uid)` method without a primaryKey, which calls `getOrCreateIndex(String uid, null)` ?

```java
public Index getOrCreateIndex(String uid) throws Exception {
	return getOrCreateIndex(uid, null);
}
```

`primaryKey` should be optional!

After this PR will be ready to merge! 🎉

**Comment by arjunrc143** [comment]
> Hey @arjunrc143!
> 
> We can already use custom Exceptions, Can you please rebase your branch on master?
> 
> I did a suggestion on your code, on how to check the `MeiliSearchApiException` errorCode in this method which is already tested!
> 
> ALSO:
> 
> Can you please create a `getOrCreateIndex(String uid)` method without a primaryKey, which calls `getOrCreateIndex(String uid, null)` ?
> 
> ```java
> public Index getOrCreateIndex(String uid) throws Exception {
> 	return getOrCreateIndex(uid, null);
> }
> ```
> 
> `primaryKey` should be optional!
> 
> After this PR will be ready to merge! 🎉

@eskombro This is done. Can you please review it?

**Review by eskombro — APPROVED** [review]
Thanks! 🎉

**Pull request merged**